// graphdb/lib/repository/repository-client-config.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

var ClientConfig = require('../http/client-config');

var RDFMimeType = require('../http/rdf-mime-type');

var defaultTimeout = 10000;
/**
 * Configuration wrapper used for initialization of {@link BaseRepositoryClient}
 * implementations.
 *
 * @class
 * @extends ClientConfig
 * @author Mihail Radkov
 * @author Svilen Velikov
 * @author Teodossi Dossev
 */

var RepositoryClientConfig = /*#__PURE__*/function (_ClientConfig) {
  _inherits(RepositoryClientConfig, _ClientConfig);

  var _super = _createSuper(RepositoryClientConfig);

  /**
   * Repository client configuration constructor.
   * Initializes [endpoints]{@link RepositoryClientConfig#endpoints} and
   * sets configuration default values to
   * [defaultRDFMimeType]{@link RepositoryClientConfig#defaultRDFMimeType},
   * [readTimeout]{@link RepositoryClientConfig#readTimeout} and
   * [writeTimeout]{@link RepositoryClientConfig#writeTimeout}
   *
   * @param {string} endpoint server base URL that will be prepend
   * to all server requests
   */
  function RepositoryClientConfig(endpoint) {
    var _this;

    _classCallCheck(this, RepositoryClientConfig);

    _this = _super.call(this, endpoint);

    _this.setEndpoints([]);

    _this.setHeaders({});

    _this.setKeepAlive(true);

    _this.setDefaultRDFMimeType(RDFMimeType.SPARQL_RESULTS_JSON);

    _this.setReadTimeout(defaultTimeout);

    _this.setWriteTimeout(defaultTimeout);

    return _this;
  }
  /**
   * Sets the repository endpoint URLs.
   *
   * @param {string[]} endpoints the endpoint URLs
   *
   * @return {this} current config for method chaining
   */


  _createClass(RepositoryClientConfig, [{
    key: "setEndpoints",
    value: function setEndpoints(endpoints) {
      this.endpoints = endpoints;
      return this;
    }
    /**
     * Inserts a repository endpoint URL to the rest of the endpoints.
     *
     * @param {string} endpoint repository endpoint URL
     *
     * @return {this} current config for method chaining
     */

  }, {
    key: "addEndpoint",
    value: function addEndpoint(endpoint) {
      if (!this.endpoints) {
        this.endpoints = [];
      }

      this.endpoints.push(endpoint);
      return this;
    }
    /**
     * Gets the repository endpoint URLs.
     *
     * @return {string[]}
     */

  }, {
    key: "getEndpoints",
    value: function getEndpoints() {
      return this.endpoints;
    }
    /**
     * Sets the default RDF MIME type.
     *
     * @param {string} defaultRDFMimeType
     *
     * @return {this} current config for method chaining
     */

  }, {
    key: "setDefaultRDFMimeType",
    value: function setDefaultRDFMimeType(defaultRDFMimeType) {
      this.defaultRDFMimeType = defaultRDFMimeType;
      return this;
    }
    /**
     * Returns the default RDF MIME type.
     *
     * @return {string}
     */

  }, {
    key: "getDefaultRDFMimeType",
    value: function getDefaultRDFMimeType() {
      return this.defaultRDFMimeType;
    }
    /**
     * Sets the default read timeout for HTTP requests.
     *
     * @param {number} readTimeout the timeout in milliseconds
     *
     * @return {this} current config for method chaining
     */

  }, {
    key: "setReadTimeout",
    value: function setReadTimeout(readTimeout) {
      this.readTimeout = readTimeout;
      return this;
    }
    /**
     * Returns the default read timeout for HTTP requests.
     *
     * @return {number}
     */

  }, {
    key: "getReadTimeout",
    value: function getReadTimeout() {
      return this.readTimeout;
    }
    /**
     * Sets the default write timeout for HTTP requests.
     *
     * @param {number} writeTimeout the timeout in milliseconds
     *
     * @return {this} current config for method chaining
     */

  }, {
    key: "setWriteTimeout",
    value: function setWriteTimeout(writeTimeout) {
      this.writeTimeout = writeTimeout;
      return this;
    }
    /**
     * Returns the default write timeout for HTTP requests.
     *
     * @return {number}
     */

  }, {
    key: "getWriteTimeout",
    value: function getWriteTimeout() {
      return this.writeTimeout;
    }
  }]);

  return RepositoryClientConfig;
}(ClientConfig);

module.exports = RepositoryClientConfig;
