// graphdb/lib/repository/statement-payload.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

/**
 * Abstract class for constructing a statement payload consisted of:
 * <ul>
 *  <li>subject</li>
 *  <li>predicate</li>
 *  <li>object</li>
 *  <li>context or contexts</li>
 * </ul>
 *
 * @class
 * @abstract
 * @author Mihail Radkov
 * @author Svilen Velikov
 */
var StatementPayload = /*#__PURE__*/function () {
  /**
   * Instantiates new statement payload.
   */
  function StatementPayload() {
    _classCallCheck(this, StatementPayload);

    this.payload = {};
  }
  /**
   * @param {string} [subject]
   * @return {StatementPayload}
   */


  _createClass(StatementPayload, [{
    key: "setSubject",
    value: function setSubject(subject) {
      this.payload.subject = subject;
      return this;
    }
    /**
     * @return {string} subject
     */

  }, {
    key: "getSubject",
    value: function getSubject() {
      return this.payload.subject;
    }
    /**
     * @param {string} [predicate]
     * @return {StatementPayload}
     */

  }, {
    key: "setPredicate",
    value: function setPredicate(predicate) {
      this.payload.predicate = predicate;
      return this;
    }
    /**
     * @return {string} predicate
     */

  }, {
    key: "getPredicate",
    value: function getPredicate() {
      return this.payload.predicate;
    }
    /**
     * @param {string} [object]
     * @return {StatementPayload}
     */

  }, {
    key: "setObject",
    value: function setObject(object) {
      this.payload.object = object;
      return this;
    }
    /**
     * @return {string} object
     */

  }, {
    key: "getObject",
    value: function getObject() {
      return this.payload.object;
    }
    /**
     * @param {(string|string[])} [context]
     * @return {StatementPayload}
     */

  }, {
    key: "setContext",
    value: function setContext(context) {
      this.payload.context = context;
      return this;
    }
    /**
     * @return {string} context
     */

  }, {
    key: "getContext",
    value: function getContext() {
      return this.payload.context;
    }
    /**
     * Get the payload object.
     * @return {Object}
     */

  }, {
    key: "get",
    value: function get() {
      return this.payload;
    }
  }]);

  return StatementPayload;
}();

module.exports = StatementPayload;
